### ISSUE TITLE
[APMMeasurement didInitializeComponments] main thread hangs

### ISSUE THREAD

**Issue by linziyiwj** [firebase/firebase-ios-sdk#10151]
### Description

Sdk version：
Firebase  9.5.0
Google-Mobile-Ads-SDK 9.8.0
GoogleAppMeasurement: Firebase/GoogleAppMeasurement

We find the problems in hangs of Xcode 14.0 organizer, it hangs the main thread for a long time.

<img width="839" alt="image" src="https://user-images.githubusercontent.com/39822433/187418273-752d0f87-a69b-4b95-9dc8-5e958b8cb0c0.png">

### Firebase SDK Version

9.5.0

### Xcode Version

13.4

### Installation Method

CocoaPods

### Firebase Product(s)

AB Testing, Analytics, Crashlytics, In-App Messaging, Performance, Remote Config

### Targeted Platforms

iOS

### If using Swift Package Manager, the project's Package.resolved

<details>
<summary>Expand <code>Package.resolved</code> snippet</summary>
<br>

```json

Replace this line with the contents of your Package.resolved.

```

</details>

### If using CocoaPods, the project's Podfile.lock

<details>
<summary>Expand <code>Podfile.lock</code> snippet</summary>
<br>

```yml

Replace this line with the contents of your Podfile.lock!

```

</details>

**Comment by rizafran** [comment]
Thanks for reaching out, @linziyiwj. Could you provide the detailed steps in reproducing the issue as well as the complete stack trace?

**Comment by morganchen12** [comment]
Looks like the lock is necessary for this initialization logic, but we could potentially speed it up by using `os_unfair_lock`.

**Comment by linziyiwj** [comment]
> 
Unfortunately I have not been able to reproduce this block yet.
It is only reported in Xcode hangs. And only the above stack information

**Comment by linziyiwj** [comment]
> Thanks for reaching out, @linziyiwj. Could you provide the detailed steps in reproducing the issue as well as the complete stack trace?

@rizafran Unfortunately I have not been able to reproduce this block yet.
It is only reported in Xcode hangs. And only the above stack information

**Comment by linziyiwj** [comment]
> Looks like the lock is necessary for this initialization logic, but we could potentially speed it up by using `os_unfair_lock`.

@morganchen12 Sorry, I don't quite understand it. Can you explain it in detail. Thank you very much!!!

**Comment by paulb777** [comment]
@linziyiwj I believe @morganchen12 was talking about a possible change in our internal implementation.

**Comment by linziyiwj** [comment]
> @linziyiwj I believe @morganchen12 was talking about a possible change in our internal implementation.

@paulb777  Please ask when will the version be updated to deal with this problem，Hope to have a solution as soon as possible

**Comment by paulb777** [comment]
cc: @htcgh

**Comment by tsunghung** [comment]
Hi @linziyiwj, could you please provide all stack traces? It would help us to know where our internal worker thread is blocked. Thanks.

**Comment by tsunghung** [comment]
@linziyiwj
Just want to be sure, it's not a dead lock, right? It's just that the main thread gets blocked for 1 second, right? Please correct me if I misunderstood.
It seems that the worker thread holds the lock while initializing internal components, and the main thread is waiting for the initialization finished.
One quick suggestion, you may want to call `FirebaseApp.configure()` as soon as possible to allow FirebaseAnalytics initialize earlier and unblock the main thread sooner.

**Comment by linziyiwj** [comment]
> It is only reported in Xcode hangs. And only the above stack information

Sorry，It is only reported in Xcode hangs. And only the above stack information

**Comment by linziyiwj** [comment]
> Hi @linziyiwj, could you please provide all stack traces, including all other threads? It would help us to know where our internal worker thread is blocked. Thanks.

@tsunghung Unfortunately I have not been able to reproduce this block yet.
It is only reported in Xcode hangs. And only the above stack information

**Comment by linziyiwj** [comment]
> @linziyiwj Just want to be sure, it's not a dead lock, right? It's just that the main thread gets blocked for 1 second, right? Please correct me if I misunderstood. It seems that the worker thread holds the lock while initializing internal components, and the main thread is waiting for the initialization finished. One quick suggestion, you may want to call `FirebaseApp.configure()` as soon as possible to allow FirebaseAnalytics initialize earlier and unblock the main thread sooner.
@tsunghung yes, It's just that the main thread gets blocked for 1 second. And We called `[FIRApp configure]` when  `- (BOOL)application:(UIApplication *)application willFinishLaunchingWithOptions:(NSDictionary<UIApplicationLaunchOptionsKey,id> *)launchOptions` on the main thread , No lock held

@linziyiwj
Author
[linziyiwj](https://github.com/linziyiwj) commented [7 minutes ago](https://github.com/firebase/firebase-ios-sdk/issues/10151#issuecomment-1236309442)
Hi @linziyiwj, could you please provide all stack traces, including all other threads? It would help us to know where our internal worker thread is blocked. Thanks.

@tsunghung Unfortunately I have not been able to reproduce this block yet.
It is only reported in Xcode hangs. And only the above stack information

@linziyiwj

 
Add heading textAdd bold text, <Cmd+b>Add italic text, <Cmd+i>

**Comment by tsunghung** [comment]
According to the call stack, we created a patch, which will be released in Firebase 10.0. Hopefully, it would fix this issue.

**Comment by ncooke3** [comment]
Hi everyone, Firebase 10 has been released and should resolve this issue. Please don't hesitate to reach out if the issue persists.
